lib/internal/loader/Translators.js

```diff
@@ -53,12 +53,31 @@ translators.set('cjs', async (url) => {
 
 // Strategy for loading a node builtin CommonJS module that isn't
 // through normal resolution
+let builtinWarned = false;
 translators.set('builtin', async (url) => {
+  if (!builtinWarned) {
+    builtinWarned = true;
+    process.emitWarning('Importing core libraries is still in development' +
+      ' and the current behavior may be changed.', 'Loader');
+  }
   debug(`Translating BuiltinModule ${url}`);
-  return createDynamicModule(['default'], url, (reflect) => {
+  const id = url.substr(5);
+  const module = NativeModule.requireNamespace(id);
+  const keys = Object.keys(module);
+  return createDynamicModule(keys, url, (reflect) => {
     debug(`Loading BuiltinModule ${url}`);
-    const exports = NativeModule.require(url.slice(5));
-    reflect.exports.default.set(exports);
+    for (const key of keys)
+      reflect.exports[key].set(module[key]);
+    reflect.exports.default.set(new Proxy(module.default, {
+      set(target, name, value) {
+        target[name] = value;
+        if (name in module)
+          module[name] = value;
+        if (name in reflect.exports)
+          reflect.exports[name].set(value);
+        return true;
+      },
+    }));
```

Comment:
I would have thought that this behaviour is exactly the same as what is possible by definining individual setters.

So it would be interesting to perhaps compare performance of setters v a proxy here to see that variation in making these decisions.

Reply:
> I would have thought that this behaviour is exactly the same as what is possible by definining individual setters.

The diff between setters and a proxy is that you can use this proxy approach on the wider CJS ecosystem since it doesn't muck with the export object. For builtins it might be less noticable since they can change all their exports to be getter/setters so that `require("fs")` and `import fs from "fs"` are `===` to each other.

Reply:
at the moment this only works one way.
```js
import fs from 'fs';

fs.readFile = 7;

// some other cjs file
require('fs').readFile === 7
```

```js
// some cjs
require('fs').readFile = 7;

// another file
import fs, { readFile } from 'fs';

readFile === 7 // not guaranteed
fs.readFile === 7 // not guaranteed
```

the second case only works if the module hasn't been inserted into the nsCache object yet. we would have to wrap the exports of all builtins with proxies or use setters/getters for every method of every builtin even in cjs to sync cjs => esm at any point during runtime

Reply:
> the second case only works if the module hasn't been inserted into the nsCache object yet. 

I don't thin so. Sine the `nsCache` will have access to the `module.exports` from `require("fs")` it should be fine even if loaded after and if cached. The `module.exports` reference itself isn't changing). You could write a test on it to verify.

Reply:
you misunderstood. i mean if you modify `module.exports` in cjs land it might not be reflected in esm land

Reply:
Oh yeah. That's fine. I don't think userland can replace `module.exports` of builtins and I'm fine with that limitation from the wider CJS ecosystem.

Reply:
i meant more like `require('any builtin').existingProperty = somethingElse` because by the time it gets modified the ns cache is already built and detatched.

Reply:
With a reference to the `module.exports` reference it shouldn't matter. The reference won't change and will give you access to properties on it for the lifetime of the reference.